### PR TITLE
DIV-5186: Create basic functional tests

### PR DESCRIPTION
**Pull request by johnoneill46**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -58,6 +59,7 @@
     "@hmcts/div-idam-test-harness": "^1.5.0",
     "@hmcts/div-test-harness": "https://github.com/hmcts/div-test-harness#0.1",
     "@hmcts/one-per-page-test-suite": "^3.0.2",
+    "@hmcts/node-js-environment-variable-setter": "https://github.com/hmcts/node-js-environment-variable-setter#0.2",
```
Comment (tchow8) [inline package.json]: Don't need this in two places.

File: package.json [hunk 2]
```diff
@@ -32,6 +32,7 @@
     "@hmcts/div-document-express-handler": "^1.1.0",
     "@hmcts/div-idam-express-middleware": "^6.4.0",
     "@hmcts/look-and-feel": "^2.2.3",
+    "@hmcts/node-js-environment-variable-setter": "https://github.com/hmcts/node-js-environment-variable-setter#0.2",
```
Comment (hosainnet) [inline package.json]: I think we moved away from using this, instead we can use the built-in `config` functionality, see the "Running tests locally against a PR/AAT" in README.md
https://github.com/hmcts/div-respondent-frontend
And https://github.com/hmcts/div-respondent-frontend/blob/master/config/example-local-aat.yml

File: test/functional/codecept.conf.js [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    IdamHelper: { require: './helpers/idamHelper.js' },
+    CaseHelper: { require: './helpers/caseHelper.js' },
+    JSWait: { require: './helpers/JSWait.js' },
+    UrlHelper: { require: './helpers/urlHelper.js' }
+  },
+  include: { I: './pages/steps.js' },
+  mocha: {
+    reporterOptions: {
+      reportDir: process.env.FUNCTIONAL_OUTPUT_DIR || './functional-output',
+      reportName: 'DecreeNisiFrontendTests',
```
Comment (tchow8) [inline test/functional/codecept.conf.js]: Change ReportName

File: test/functional/codecept.smoke.conf.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+  output: config.tests.functional.outputDir,
+  helpers: {
+    Puppeteer: {
+      url: config.tests.functional.url || config.node.baseUrl,
+      waitForTimeout,
+      waitForAction,
+      show: false,
+      restart: false,
+      keepCookies: false,
+      keepBrowserState: false,
```
Comment (hosainnet) [inline test/functional/codecept.smoke.conf.js]: We should remove 
```

      restart: false,
      keepCookies: false,
      keepBrowserState: false,
```
They make the tests fail on non-AKS - me and @tchow8  talked about it a few days ago (removed it from Demo/ITHC branches of DN)

File: test/functional/codecept.smoke.conf.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+exports.config = {
+  tests: './smoke/*.js',
+  output: config.tests.functional.outputDir,
+  helpers: {
+    Puppeteer: {
+      url: config.tests.functional.url || config.node.baseUrl,
+      waitForTimeout,
+      waitForAction,
+      show: false,
+      restart: false,
```
Comment (tchow8) [inline test/functional/codecept.smoke.conf.js]: Remove restart, keepCookies and keepBrowserState. They cause an issue on Demo and ITHC and aren't needed.

File: test/functional/helpers/idamHelper.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+
+class IdamHelper extends Helper {
+  createAUser() {
+    if (parseBool(config.features.idam)) {
+      const randomString = randomstring.generate({
+        length: 16,
+        charset: 'numeric'
+      });
+      const emailName = `divorce+dn-test-${randomString}`;
+      const testEmail = `${emailName}@example.com`;
```
Comment (hosainnet) [inline test/functional/helpers/idamHelper.js]: I know this probably exists in the other projects, but we should maybe use `@localhost.local` domain or something, as `example.com` could have a "catch-all" setup which means they'd receive the emails

File: test/functional/saucelabs.conf.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+const browser = process.env.SAUCE_BROWSER || config.saucelabs.browser;
+const tunnelName = process.env.SAUCE_TUNNEL_IDENTIFIER || config.saucelabs.tunnelId;
+const getBrowserConfig = browserGroup => {
+  const browserConfig = [];
+  for (const candidateBrowser in supportedBrowsers[browserGroup]) {
+    if (candidateBrowser) {
+      const desiredCapability = supportedBrowsers[browserGroup][candidateBrowser];
+      desiredCapability.tunnelIdentifier = tunnelName;
+      desiredCapability.tags = ['DN_divorce'];
```
Comment (tchow8) [inline test/functional/saucelabs.conf.js]: Change to DA

File: test/functional/saucelabs.conf.js [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+      browsers: getBrowserConfig('chrome')
+    },
+    firefox: {
+      browsers: getBrowserConfig('firefox')
+    },
+    safari: {
+      browsers: getBrowserConfig('safari')
+    }
+  },
+  name: 'DN Frontend Tests'
```
Comment (tchow8) [inline test/functional/saucelabs.conf.js]: Change to DA

File: test/resources/errors.js [hunk 9]
```diff
@@ -0,0 +1,31 @@
+const httpStatus = require('http-status-codes');
```
Comment (tchow8) [inline test/resources/errors.js]: Can remove this file and the img files right? No upload mechanism in DA.

File: test/functional/codecept.smoke.conf.js (hunk 5 above)
Comment (tchow8) [reply test/functional/codecept.smoke.conf.js]: Hosain raises this below.

**Pull request merged**